ciphey/basemods/Crackers/XandY.py

```diff
@@ -0,0 +1,108 @@
+# community
+# by https://github.com/lukasgabriel
+
+from distutils import util
+from typing import Optional, Dict, Union, Set, List
+from loguru import logger
+
+from ciphey.iface import ParamSpec, Cracker, CrackResult, CrackInfo, T, registry, Config
+
+
+@registry.register
+class XandY(Cracker[str]):
+    def getInfo(self, ctext: str) -> CrackInfo:
+        # TODO Write something useful here
+        return CrackInfo(
+            success_likelihood=1, success_runtime=1e-5, failure_runtime=1e-5,
+        )
+
+    @staticmethod
+    def binary_to_ascii(variant):
+        # Convert the binary string to an integer with base 2
+        binary_int = int(variant, 2)
+        byte_number = binary_int.bit_length() + 7 // 8
+
+        # Convert the resulting int to a bytearray and then decode it to ascii text
+        binary_array = binary_int.to_bytes(byte_number, "big")
+        try:
+            ascii_text = binary_array.decode()
+            logger.trace(f"Found possible solution: {ascii_text[:32]}...")
+            return ascii_text
+        except UnicodeDecodeError:
+            return ""
+
+    @staticmethod
+    def getTarget() -> str:
+        return "XandY"
+
+    def attemptCrack(self, ctext: str) -> List[CrackResult]:
+        """
+        Checks an input if it only consists of two different letters.
+        If this is the case, it attempts to regard those two letters as
+        0 and 1 and then converts it to ascii text.
+        """
+        logger.trace("Attempting X-Y replacement.")
+        variants = []
+        candidates = []
+        result = []
+        ctext = ctext.lower().replace(" ", "").strip()
+
+        # cset contains every unique value in the cstring
+        cset = list(set(list(ctext)))
+        if len(cset) != 2:
+            # We only consider inputs with exactly two unique values
+            logger.trace("String contains more than two unique values. Skipping X-Y...")
+            return None
+        else:
+            logger.trace(f"String contains two unique values: {cset[0], cset[1]}")
+            # Form both variants of the substitution
+            for i in range(2):
```

Comment:
This is making the assumption that the string only contains 2 chars, which is wrong I believe. Binary strings in CTFs often have spaces, commas, new lines to delimiter themselves:
https://gchq.github.io/CyberChef/#recipe=To_Binary('Space')&input=aGVsbG8gbXkgbmFtZSBpcw

Reply:
when i say i believe, i mean i believe this is making that assumption

Reply:
If I was you, I would make this a 3 way checker, where the 3rd check is a delimiter for a space, however.... this is optional, because Binary also natively supports spaces in the form of ASCII :P

Reply:
But if it contains 3 unique chars, it might be using a space -- whereas 2 unique chars we know it can't be (unless its a binary string of all 1s)

Reply:
Spaces are removed in line 47; as for other delimiters, I'd have to go back and think about how to check for that. After all, the delimiters could themselves be representing 1s or 0s (in theory).

Reply:
Now I get what you mean by three-way check for spaces :p  Yeah, that would be an idea. But I feel like I'm in over my head already

Reply:
Honestly the way I would do it would be the way you're doing it but adding in another one.

Imo, if it's hard to do this then the function isn't right -- it should theoretically be possible to construct a single function which operates on X and Y, X and Y and P, and so on ad infinitum

Is there anyway you can make a single function? Abstract away the code? :) 

If you had to make a function to do this, but you got told you didn't know how many chars were in the input, how would you abstract away hardcoded values into a one-size-fits-all solution? :) 

Even if you gave rough psuedocode, I can help you fill it in :)